pipeline_branching.md

```diff
@@ -0,0 +1,159 @@
+just stick to that for inspiration.
+
+## PPS API changes:
+The pipeline message will be extended to include a `Branch` field, which
+is a string:
+
+```proto
+message Pipeline {
+  string name = 1;
+  string branch = 2;
+}
+```
+
+In pachctl this is implemented with the familiar syntax of `pipeline@branch`.
+
+An empty `branch` field is equivalent of a value of `"master"`.
+
+Different branches of the same pipeline are effectively different pipelines. That means:
+- They output to different branches, with names that match their branch names.
+- You can delete, update, pause, etc. one without pausing the other.
+- They can take different inputs, have different resource requests, etc.
+- They run in different pods, and their success / failure is independent.
+
+There are a few ways in which they're not quite totally distinct:
+- They consolidate in list pipeline, this one's mostly cosmetic, I'm
+imagining something that looks like this:
+```
+NAME     BRANCHES     INPUT    CREATED    STATE / LAST JOB  DESCRIPTION
+pipeline dev, staging input:/* 6 days ago running / success
+```
+- They share datums for deduplication, if you want a dev version to
+reprocess, you need to create it with `--reprocess`.
+- They can't take each other as input. Implementation wise there's no
+particular reason for this, but it seems very weird and I can't think of any
+reason you'd want to do this that there isn't a better way of accomplishing.
+
+### A few examples of what syntax would look like for this:
+Creating a dev branch:
+
+```
+extract pipeline pipeline >pipeline.json
+# modify pipeline.json, setting Pipeline.Branch to "dev"
+create pipeline -f pipeline.json
+```
+
+List jobs from dev branch:
+
+```
+list job -p pipeline@dev
+```
+
+Get files from dev branch (all other file commands work similarly):
+
+```
+get file pipeline@dev:/file
+```
+
+Promote dev branch to master
+
+```
+# modify pipeline.json, setting Pipeline.Branch to ""
+create pipeline -f pipeline.json
+```
+
+## Downstream Pipelines:
+The above changes are fairly straightforward, both in terms of how we
+would implement them and in terms of explaining them to users. Downstream
+pipelines is where it starts to get complicated. Downstream pipelines are
+important to consider because when you're iterating on a pipeline, you
+care that the results of your new version can be consumed by the rest of
+the DAG downstream. For example, suppose I have the pipelines: A, B and C
+with a DAG like this:
+
+```
+A -> B -> C
+```
+
+Initially, each pipeline has only a `master` version, then we create
+`A@dev` this should create branches of `B` and `C` also called, `dev`
+which process `A@dev` and `B@dev` respectively. So now the picture looks
+like this:
+
+```
+A -> B -> C
+A@dev -> B@dev -> C@dev
+```
+
+B@dev and C@dev are identical to B and C, except that they process the
+`dev` branch of their inputs, rather than `master`.
+
+Adding a new pipeline works similarly, if someone comes and creates `D`
```

Comment:
Since the user can update the @dev what's expected to happens if they add a new pipeline to @dev?

Reply:
I think you're asking what happens if they create `D`, which automatically creates `D@dev` and then they later want to modify `D@dev`. This is completely allowed, users can modify `D@dev` however they want. Generally speaking I expect that `dev` will wind up being a mix of modified steps in the places you want to modify the dag, and unmodified steps that you want to have there to make sure your modified steps aren't going to break downstream things.

Reply:
More clearly, can a user create D only in @dev and not in @master?

Reply:
Oh I see, I hadn't thought of that as a use case, I think it technically should work, although it seems a little bit weird to me to have a dev version of the pipeline, without a master version, seems like you might as well just call it master, since you're not using that name for anything else.

Reply:
The flow I was thinking... the is a "prod" version of the DAG. Customer create a "dev" branch, they extend the DAG with additional pipelines steps, test/experiment in the "dev" branch. eventually they promote it to master. At that point, master now has a new DAG